[ActiveStorage] Prevent `AS::Preview#processed` to generate a variant for an empty transformation

chaadow opened the pull request:
Followup #50044 cc @jonathanhefner 


<!--
Thanks for contributing to Rails!

Please do not make *Draft* pull requests, as they still send
notifications to everyone watching the Rails repo.

Create a pull request when it is ready for review and feedback
from the Rails team :).

If your pull request affects documentation or any non-code
changes, guidelines for those changes are [available
here](https://edgeguides.rubyonrails.org/contributing_to_ruby_on_rails.html#contributing-to-the-rails-documentation)

About this template

The following template aims to help contributors write a good description for their pull requests.
We'd like you to provide a description of the changes in your pull request (i.e. bugs fixed or features added), motivation behind the changes, and complete the checklist below before opening a pull request.

Feel free to discard it if you need to (e.g. when you just fix a typo). -->

### Motivation / Background

<!--
Describe why this Pull Request needs to be merged. What bug have you fixed? What feature have you added? Why is it important?
If you are fixing a specific issue, include "Fixes #ISSUE" (replace with the issue number, remove the quotes) and the issue will be linked to this PR.
-->


if an empty hash is passed to a preview call (`blob.preview({})`) We go through the original preview instead of regenerating a variation based on the original preview image which would result in a performance penalty.

### Detail

Calls to `#url` or `#key` would now use the original `blob.preview_image` instead of generating a variant that has an empty transformation.

==> Resulting in saving both a `MiniMagic/vips` call. + a network/IO call to the storage service

### Additional information

<!-- Provide additional information such as benchmarks, reference to other repositories or alternative solutions. -->

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Changes that are unrelated should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [x] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

chaadow commented on activestorage/test/models/preview_test.rb:
```diff
@@ -70,6 +70,50 @@ class ActiveStorage::PreviewTest < ActiveSupport::TestCase
       preview.processed
     end
+
+    image = read_image( preview.image.variant(transformations))
```
@jonathanhefner i've added a test to followup on your PR here.

also below,  i've added basic tests for all of `#download` `#key` and `#url` since they were missing and they're public API.

chaadow committed:
Prevent `AS::Preview#processed` to generate an empty variant

if an empty hash is passed to a preview call (`blob.preview({})`)
We go through the original preview instead of regenerating a variation
based on the original preview image which would result in a performance
penalty

jonathanhefner commented on activestorage/test/models/preview_test.rb:
```diff
@@ -70,6 +70,61 @@ class ActiveStorage::PreviewTest < ActiveSupport::TestCase
       preview.processed
     end
+
+    image = read_image(preview.image.variant(transformations))
+    assert_equal 216, image.width
+    assert_equal 280, image.height
```
I don't think these assertions are necessary for this particular test.  This test only cares about whether the variant was processed.  Tests in `variant_test.rb` should check whether `Variant#processed` works as expected.

chaadow replied:
I agree

jonathanhefner commented on activestorage/test/models/preview_test.rb:
```diff
@@ -70,6 +70,61 @@ class ActiveStorage::PreviewTest < ActiveSupport::TestCase
+    assert_equal 280, image.height
+  end
+
+  test "#processed does not generate a variant when provided an empty transformation" do
+    blob = create_file_blob(filename: "report.pdf", content_type: "application/pdf")
+    empty_transformation = {}
+    preview = blob.preview(empty_transformation)
+
+    assert_no_changes -> { !!preview.image.variant(empty_transformation)&.send(:processed?) } do
+      preview.processed
+    end
+  end
+
+  test "#download does not generate a variant when provided an empty transformation" do
+    blob = create_file_blob(filename: "report.pdf", content_type: "application/pdf")
+    empty_transformation = {}
+    preview = blob.preview(empty_transformation)
+
+    assert_no_changes -> { !!preview.image.variant(empty_transformation)&.send(:processed?) } do
+      preview.processed.download {  }
+    end
+  end
+
+  test "#key does not generate a variant when provided an empty transformation" do
+    blob = create_file_blob(filename: "report.pdf", content_type: "application/pdf")
+    empty_transformation = {}
+    preview = blob.preview(empty_transformation)
+
+    assert_no_changes -> { !!preview.image.variant(empty_transformation)&.send(:processed?) } do
+      preview.processed.key
+    end
+  end
```
I think we can combine these tests (and the `"#url does not generate a variant when provided an empty transformation"` test further below) into a single `"previewing with empty transformations does not generate a variant"` test.

chaadow replied:
https://github.com/rails/rails/pull/50046#discussion_r1393143712

jonathanhefner commented on activestorage/test/models/preview_test.rb:
```diff
@@ -70,6 +70,61 @@ class ActiveStorage::PreviewTest < ActiveSupport::TestCase
+
+  test "#processed does not generate a variant when provided an empty transformation" do
+    blob = create_file_blob(filename: "report.pdf", content_type: "application/pdf")
+    empty_transformation = {}
+    preview = blob.preview(empty_transformation)
```
```suggestion
    preview = blob.preview({})
```

jonathanhefner commented on activestorage/test/models/preview_test.rb:
```diff
@@ -70,6 +70,61 @@ class ActiveStorage::PreviewTest < ActiveSupport::TestCase
+    empty_transformation = {}
+    preview = blob.preview(empty_transformation)
+
+    assert_no_changes -> { !!preview.image.variant(empty_transformation)&.send(:processed?) } do
+      preview.processed
+    end
```
Instead of `assert_no_changes`, let's just assert that there are no variant records after all is done:

```suggestion
    preview.processed
    # etc

    assert_empty preview.image.variant_records
```

jonathanhefner commented on activestorage/test/models/preview_test.rb:
```diff
@@ -70,6 +70,61 @@ class ActiveStorage::PreviewTest < ActiveSupport::TestCase
+    preview = blob.preview(empty_transformation)
+
+    assert_no_changes -> { !!preview.image.variant(empty_transformation)&.send(:processed?) } do
+      preview.processed.key
```
It might be good to also assert that `preview.key` is equal to `blob.preview_image.key`.

jonathanhefner commented on activestorage/test/models/preview_test.rb:
```diff
@@ -70,6 +70,61 @@ class ActiveStorage::PreviewTest < ActiveSupport::TestCase
+    preview = blob.preview(empty_transformation)
+
+    assert_no_changes -> { !!preview.image.variant(empty_transformation)&.send(:processed?) } do
+      preview.processed.url
```
It might be good to also assert that the `preview.url` is equal to `blob.preview_image.url`.

chaadow replied:
I agree

jonathanhefner commented on activestorage/test/models/preview_test.rb:
```diff
@@ -70,6 +70,61 @@ class ActiveStorage::PreviewTest < ActiveSupport::TestCase
+    end
+  end
+
+  test "#key #download and #url raise `UnprocessedError` when preview is not processed" do
+    blob = create_file_blob(filename: "report.pdf", content_type: "application/pdf")
+    preview = blob.preview({})
```
Since this behavior isn't specific to empty transformations, and specifying non-empty transformations is the common case, let's do that here:

```suggestion
  test "methods raise UnprocessedError when preview is not processed" do
    blob = create_file_blob(filename: "report.pdf", content_type: "application/pdf")
    preview = blob.preview(resize_to_limit: [640, 280])
```

chaadow replied:
`methods raise ...` I find the "methods" very generic, I'm going to suggest a slight refactoring, and let me know what you think.

I agree with the non empty transformations 👍🏼

jonathanhefner commented on activestorage/CHANGELOG.md:
```diff
@@ -1,3 +1,9 @@
+*   Prevent `ActiveStorage::Blob#preview` to generate a variant if an empty variation is passed.
+    calls to `#url`, `#key` or `#download` will now use the original preview
```
```suggestion
    Calls to `#url`, `#key` or `#download` will now use the original preview
```

chaadow commented on activestorage/app/models/active_storage/preview.rb:
```diff
@@ -105,14 +105,17 @@ def process
+      image.variant(variation)
     end
 
-
```
Yeah sorry it's my editor(vim) :/ i'll try to put it back

chaadow replied:
Done

chaadow commented on activestorage/app/models/active_storage/preview.rb:
```diff
@@ -63,7 +63,7 @@ def image
   def url(**options)
     if processed?
-      variant.url(**options)
+      empty_variation? ? image.url(**options) : variant.processed.url(**options)
```
I love this

chaadow commented on activestorage/test/models/preview_test.rb:
```diff
@@ -4,6 +4,8 @@
 require "database/setup"
 
 class ActiveStorage::PreviewTest < ActiveSupport::TestCase
+  PREVIEW_METHODS = %i[key url download]
```
@jonathanhefner Since these methods are public method and public interface as well, I wanted to test them each individually even though they are similar.

I believe in the future, adding more public methods to AS:Preview would simply require adding the method under test to this constant.

Right now I believe I'm testing 2/3 of these methods. the 1/3 left is when provided a non blank transformation. I don't know if you think it's worth testing.

chaadow replied:
I decided to add the missing third. And with this, I believe we have full coverage of all of the three preview methods.

chaadow commented on activestorage/test/models/preview_test.rb:
```diff
@@ -72,6 +74,41 @@ class ActiveStorage::PreviewTest < ActiveSupport::TestCase
+
+  PREVIEW_METHODS.each do |method_under_test|
+    test "##{method_under_test} does not generate a variant when provided an empty transformation" do
+      freeze_time do
```
I added `freeze_time` to handle random failure tests with calls to `#url`

chaadow commented:
I've squashed and rebased with main.

jonathanhefner commented on activestorage/test/models/preview_test.rb:
```diff
@@ -72,6 +74,56 @@ class ActiveStorage::PreviewTest < ActiveSupport::TestCase
+  PREVIEW_METHODS.each do |method_under_test|
+    test "##{method_under_test} does not generate a variant when provided an empty transformation" do
+      freeze_time do
+        blob = create_file_blob(filename: "report.pdf", content_type: "application/pdf")
+        empty_transformation = {}
+        preview = blob.preview(empty_transformation)
+
+        assert_no_changes -> { !!preview.image.variant(empty_transformation)&.send(:processed?) } do
+          preview.processed.send(method_under_test)
+        end
+        assert_equal blob.preview_image.send(method_under_test), preview.processed.send(method_under_test)
+      end
+    end
+  end
+
+  PREVIEW_METHODS.each do |method_under_test|
+    test "##{method_under_test} returns the variant's #{method_under_test} when a transformation is present" do
+      freeze_time do
+        blob = create_file_blob(filename: "report.pdf", content_type: "application/pdf")
+        preview = blob.preview(resize_to_limit: [640, 280])
+        preview.processed
+        expected_result = preview.image.variant_records[0].image.send(method_under_test)
+
+        actual_result = preview.send(method_under_test)
+
+        assert_equal expected_result, actual_result
+      end
+    end
+  end
+
+  PREVIEW_METHODS.each do |method_under_test|
+    test "#{method_under_test} raises `UnprocessedError` when preview is not processed" do
+      blob = create_file_blob(filename: "report.pdf", content_type: "application/pdf")
+      preview = blob.preview(resize_to_limit: [640, 280])
+
+      assert_raises ActiveStorage::Preview::UnprocessedError do
+        preview.send(method_under_test)
+      end
+    end
+  end
```
Declaring tests in a loop like this can be a good strategy for exhaustive test coverage.  However, in this case, I think it adds unnecessary complexity.  It is also a leaky abstraction.  `PREVIEW_METHODS` does not accommodate methods that require arguments.  So if we were to add such methods in the future, we would need to rework this entire approach.

I believe we can cover all the behavior we're interested in with the following two tests:

```ruby
test "image-related methods raise UnprocessedError when preview is not processed" do
  blob = create_file_blob(filename: "report.pdf", content_type: "application/pdf")
  preview = blob.preview(resize_to_limit: [640, 280])

  assert_raises(ActiveStorage::Preview::UnprocessedError) { preview.url }
  assert_raises(ActiveStorage::Preview::UnprocessedError) { preview.key }
  assert_raises(ActiveStorage::Preview::UnprocessedError) { preview.download }
end

test "previewing with empty transformations does not generate a variant" do
  blob = create_file_blob(filename: "report.pdf", content_type: "application/pdf")
  preview = blob.preview({})

  preview.processed

  freeze_time { assert_equal blob.preview_image.url, preview.url }
  assert_equal blob.preview_image.key, preview.key
  assert_equal blob.preview_image.download, preview.download

  assert_empty preview.image.variant_records
end
```

Let's use those instead.

chaadow replied:
Done. thanks

jonathanhefner reviewed (commented):
I see that you recently pushed many documentation changes.  Would you mind splitting those out into separate PRs unless they are directly related to this change?  It makes things easier to review.  Also, keep in mind that this is an optimization for a corner case, so I don't think we need to extensively document it (if at all).

chaadow commented:
> Would you mind splitting those out into separate PRs unless they are directly related to this change?

@jonathanhefner I changed all of `processed`, `url`, `key` and `download`, and they're publicly documented API, I thought I would update the documentation to reflect the new reality of these methods. Even though it's an "edge case", I thought documentation should elicit all possible cases.

Moreover, contextually speaking, since it's a followup from our past Pull request that we co-authored, in which we changed `#processed` but we forgot to update the documentation, (and I also re-changed this method again), I thought I might update to further help the readers with multiple examples that could potentially help them by adding more context.

I will revert and create another PR. 

Thanks again for the prompt and exhaustive review!

chaadow commented:
@jonathanhefner I've applied the last changes, squashed the commits, and rebased with main.

jonathanhefner commented:
Thank you, @chaadow! :heavy_check_mark:

nicowenterodt commented:
Thanks from me as well. Very useful 👏 @chaadow @jonathanhefner

zzak commented on activestorage/test/models/preview_test.rb:
```diff
@@ -72,6 +72,27 @@ class ActiveStorage::PreviewTest < ActiveSupport::TestCase
+
+    preview.processed
+
+    freeze_time { assert_equal blob.preview_image.url, preview.url }
```
@chaadow I've stumbled upon this now, why is this here? If I remove it the tests still pass for me.
